### ISSUE TITLE
Bundled 7z broken on alpine

### ISSUE THREAD

**Issue by fredrikekre** [JuliaLang/julia#39154]
```
Error loading shared library libstdc++.so.6: No such file or directory (needed by /root/julia-1.6.0-beta1/libexec/7z)
Error loading shared library libgcc_s.so.1: No such file or directory (needed by /root/julia-1.6.0-beta1/libexec/7z)
Error relocating /root/julia-1.6.0-beta1/libexec/7z: _ZdlPv: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: _Znam: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: __cxa_end_catch: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: __cxa_allocate_exception: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: __cxa_pure_virtual: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: _ZdaPv: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: __gxx_personality_v0: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: __cxa_begin_catch: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: __cxa_rethrow: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: __cxa_throw: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: _Znwm: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: _Unwind_Resume: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: _ZSt9terminatev: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: __cxa_call_unexpected: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: __cxa_free_exception: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: _ZTIPKc: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: _ZTVN10__cxxabiv116__enum_type_infoE: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: _ZTIPKw: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: _ZTIi: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: _ZTVN10__cxxabiv120__si_class_type_infoE: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: _ZTVN10__cxxabiv117__class_type_infoE: symbol not found
Error relocating /root/julia-1.6.0-beta1/libexec/7z: _ZTVN10__cxxabiv121__vmi_class_type_infoE: symbol not found
```
see https://github.com/docker-library/julia/pull/50, https://github.com/docker-library/julia/pull/51

cc @tianon

**Comment by staticfloat** [comment]
This sounds like the docker image is trying to run a `glibc` binary on a `musl` system; which Julia archive is being downloaded?

**Comment by tianon** [comment]
This is from https://julialang-s3.julialang.org/bin/musl/x64/1.6/julia-1.6.0-beta1-musl-x86_64.tar.gz :sweat_smile:

**Comment by tianon** [comment]
See https://github.com/docker-library/julia/pull/51/checks?check_run_id=1671360232#step:5:76 for where it gets downloaded and https://github.com/docker-library/julia/pull/51/checks?check_run_id=1671360232#step:7:12 for where it ends up invoking `/usr/local/julia/libexec/7z` (and failing).

**Comment by tianon** [comment]
Easy local reproducer if you've got Docker (where you can thus watch it download the `musl` tarball):

```console
$ docker build 'https://github.com/docker-library/julia.git#refs/pull/51/merge:1.6-rc/alpine3.12' --pull --tag julia:1.6.0-beta1-alpine3.12
...
$ docker run --rm julia:1.6.0-beta1-alpine3.12 /usr/local/julia/libexec/7z --help
Error loading shared library libstdc++.so.6: No such file or directory (needed by /usr/local/julia/libexec/7z)
Error loading shared library libgcc_s.so.1: No such file or directory (needed by /usr/local/julia/libexec/7z)
Error relocating /usr/local/julia/libexec/7z: _ZdlPv: symbol not found
Error relocating /usr/local/julia/libexec/7z: _Znam: symbol not found
Error relocating /usr/local/julia/libexec/7z: __cxa_end_catch: symbol not found
Error relocating /usr/local/julia/libexec/7z: __cxa_allocate_exception: symbol not found
Error relocating /usr/local/julia/libexec/7z: __cxa_pure_virtual: symbol not found
Error relocating /usr/local/julia/libexec/7z: _ZdaPv: symbol not found
Error relocating /usr/local/julia/libexec/7z: __gxx_personality_v0: symbol not found
Error relocating /usr/local/julia/libexec/7z: __cxa_begin_catch: symbol not found
Error relocating /usr/local/julia/libexec/7z: __cxa_rethrow: symbol not found
Error relocating /usr/local/julia/libexec/7z: __cxa_throw: symbol not found
Error relocating /usr/local/julia/libexec/7z: _Znwm: symbol not found
Error relocating /usr/local/julia/libexec/7z: _Unwind_Resume: symbol not found
Error relocating /usr/local/julia/libexec/7z: _ZSt9terminatev: symbol not found
Error relocating /usr/local/julia/libexec/7z: __cxa_call_unexpected: symbol not found
Error relocating /usr/local/julia/libexec/7z: __cxa_free_exception: symbol not found
Error relocating /usr/local/julia/libexec/7z: _ZTIPKc: symbol not found
Error relocating /usr/local/julia/libexec/7z: _ZTVN10__cxxabiv116__enum_type_infoE: symbol not found
Error relocating /usr/local/julia/libexec/7z: _ZTIPKw: symbol not found
Error relocating /usr/local/julia/libexec/7z: _ZTIi: symbol not found
Error relocating /usr/local/julia/libexec/7z: _ZTVN10__cxxabiv120__si_class_type_infoE: symbol not found
Error relocating /usr/local/julia/libexec/7z: _ZTVN10__cxxabiv117__class_type_infoE: symbol not found
Error relocating /usr/local/julia/libexec/7z: _ZTVN10__cxxabiv121__vmi_class_type_infoE: symbol not found
```

**Comment by staticfloat** [comment]
Ah, okay, I see the issue.  The `7z` binary relies upon `libstdc++.so.6` which we ship in the `lib/julia` directory, but that directory is (obviously) not on the default library search path.  This is intentional, as we don't want to interfere with other packages that may not want our version of `libstdc++` or other libraries.

All invocations of this `7z` binary are expected to happen through Julia code, where we will have already set up `LD_LIBRARY_PATH` and such so that the executable is able to be executed properly, only that setup isn't happening properly.  This is what we get for not having any developers that use `musl`-based distros, I suppose.  I'll submit a fix for this, but we may have to live without an alpine tag on the docker image until beta2.

**Comment by staticfloat** [comment]
PR here: https://github.com/JuliaLang/julia/issues/39154

**Comment by fredrikekre** [comment]
#39155

**Comment by tianon** [comment]
I can see that this fix was backported in 1.6-rc1, but I'm still seeing the same issue (https://github.com/docker-library/julia/runs/1856690398?check_suite_focus=true#step:7:12 from https://github.com/docker-library/julia/pull/51). :confused:

Is it possible the fix was incomplete or that it conflicted with another change somehow?

**Comment by staticfloat** [comment]
It appears that the fix was incomplete; The way that `Pkg` invokes `7z` dodges the environment variables that I added in `Base`.

One workaround: define `LD_LIBRARY_PATH` in the docker container to point to `julia-1.6.0-rc1/lib/julia` (gross).  I will open a PR against `Pkg.jl` to fix the invocation, but of course it's not available in `rc1`.
